### PR TITLE
radiff2 speedups

### PR DESCRIPTION
**Pull request by NikolaiHampton**
Speeds up the radiff2 statistical diff. Drastically reduced the ‘edit
distance search space’ by ignoring values that can’t affect the edit
distance.

Testing on MacBook has a ~28 hour file diff operation down to ~13 minutes.

### CODE REVIEW

**Comment by NikolaiHampton** [comment]
So it has been a bit of a journey with some test code and trying to over-optimise some strange edge conditions. But this is the best I can do at this point in time.
I am testing a bunch of files for statistical differences, and the time complexity for the original Levenshtein distance calculation was O(NM). For similar files we can greatly reduce that by reducing the search space.
Some files speed up 10-100X others only a few times faster, but it's still an improvement.
I think we could do more to optimise some conditions, however I was having issues with special cases on edge conditions. It's easier to lose some efficiency than try to pick up rare cases.

**Comment by NikolaiHampton** [comment]
screen shot 2016-07-08 at 3 16 54 pm
The screenshot from an excel spreadsheet shows how much the search space can be reduced.
The 'red' squares are shortest paths, the bold borders are the calculated values.
Yellow squares are a special edge case for horizontal or vertical extension.
The more similar the file, the narrow that search area is.

File: libr/util/diff.c [hunk 1]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
+	// inspected all the bytes of file B..
+	//  AAAA
+	// B*
+	// B *
+	// B  *
+	// B   *
+	// B   ?
+
+	if (la < lb) {
+		aBufPtr=b;
```
Comment (radare) [inline libr/util/diff.c]: missing spaces

File: libr/util/diff.c [hunk 2]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
+
+	/* TODO: ensure those pointers are allocated */
+	int *v0 = (int*) calloc ((bLen + 3), sizeof (int));
+	int *v1 = (int*) calloc ((bLen + 3), sizeof (int));
+	int *temp;
+
+
+
 	if (!a || !b || la < 1 || lb < 1) {
+		free (v0);
```
Comment (radare) [inline libr/util/diff.c]: missing spaces

File: libr/util/diff.c [hunk 3]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
-		currentDistance[start]=previousDistance[start]+1;
-		int columnMinimum=la;
-		int stop=(i + extension + 1) < bLen ? (i + extension + 1) : bLen;
+		stop = R_MIN( (i+extendStop+2), bLen);
+		// We need a value in the result column (v1[start]).
+		// If you look at the loop below, we need it because we look at v1[j] as one of the
+		// potential shortest edit distances.
+		// In all cases where the edit distance can't 'reach',
+		// the value of v1[start] simply increments.
+        v1[start]=v0[start]+1;
```
Comment (radare) [inline libr/util/diff.c]: wrong indent

File: libr/util/diff.c [hunk 4]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
 	int start=0;
-	int extension=0;
-	int smallest;
-	int cost;
+	int stop;
+	int colMin=0;
+	// loops can get very big, this can be remove, but it'l currently in there for debugging
+	// and optimisation testing.
+	uintmax_t loops=0;
+	int extendStop=0;
```
Comment (radare) [inline libr/util/diff.c]: spacing around =

File: libr/util/diff.c [hunk 5]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
+	// that require us to keep some more data.
 	int start=0;
-	int extension=0;
-	int smallest;
-	int cost;
+	int stop;
+	int colMin=0;
+	// loops can get very big, this can be remove, but it'l currently in there for debugging
+	// and optimisation testing.
+	uintmax_t loops=0;
```
Comment (radare) [inline libr/util/diff.c]: what about using ut64 instead? uintmax_t is not used anywehre in r2land

File: libr/util/diff.c [hunk 6]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
-		// If the last numbers we looked at match we have an insertion so
-		// extend our search below the diagonal
-		if (!cost) {
-			extension++;
+		// We're going to start at i+1 next iteration
+		// The column minimum is the current edit distance
+		// This distance is the minimum 'search width' from the optimal 'i' diagonal
+		// The extendStart picks up an edge case where we have a match on the first iteration
+		// We update extendStart after we've set start for the next iteration.
+        start=i+1-colMin-extendStart;
```
Comment (radare) [inline libr/util/diff.c]: spacings

File: libr/util/diff.c [hunk 7]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
+            // so don't increment the start counter this time, BUT keep
+        	// extendStart up our sleeves for next iteration.
+            start --;
+            extendStart++;
+        }
+		temp=v0;
+		v0=v1;
+		v1=temp;
+
+		if (verbose && i % 10000==0) {
```
Comment (radare) [inline libr/util/diff.c]: add parenthesis

File: libr/util/diff.c [hunk 8]
```diff
@@ -198,165 +198,163 @@ R_API int r_diff_buffers(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb)
...
-			if (i % 10000 == 0)
-				eprintf ("Processing %d of %d (search width %i)\r", i, aLen,stop-start);
+        if (a[i+1]==b[start]) {
+            // If new start would be a match then we have a horizontal 'double up'
+        	// which means we need to keep an extra row of data
+            // so don't increment the start counter this time, BUT keep
+        	// extendStart up our sleeves for next iteration.
+            start --;
+            extendStart++;
+        }
```
Comment (radare) [inline libr/util/diff.c]: wrongly indented (space vs tabs) = use only tabs

**Comment by radare** [comment]
Can you squash the commits for proper gh review? the changes look good indeed, but i would like to have all those indentation bits fixed before merging. And check if valgrind and gcc warnings are happy.

Impressive optimizations. Thanks!

**Comment by NikolaiHampton** [comment]
Sure I'll have a look at Valgrind. There are no GCC warnings.
Don't expect this before the end of the week though. I have a paper to write :)

**Comment by alvarofe** [comment]
Just testing with ASAN

```
air:r2 alvaro$ radiff2 -g main /usr/bin/true /usr/bin/false
=================================================================
==90748==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200004c59b at pc 0x000106d0147c bp 0x7fff5b68c8f0 sp 0x7fff5b68c8e8
READ of size 1 at 0x60200004c59b thread T0
    #0 0x106d0147b in r_diff_buffers_distance diff.c:305
    #1 0x104c0a82f in r_anal_diff_fcn diff.c:214
    #2 0x1047877bb in r_core_gdiff gdiff.c:54
    #3 0x104574154 in main radiff2.c:422
    #4 0x7fff9b1db5ac in start (libdyld.dylib+0x35ac)
    #5 0x4  (<unknown module>)

ASAN:DEADLYSIGNAL
==90748==AddressSanitizer: while reporting a bug found another one. Ignoring
```

**Comment by NikolaiHampton** [comment]
I've added the calloc() pointer checks, and I'll neaten the code and squash the commits tomorrow, then I'm out of time again.

- I have no GCC warnings.
- valgrind indicates a leak. BUT, I commented out the whole r_diff_buffers_distance with a simple return, the leak is the same size:

```
R_API bool r_diff_buffers_distance(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb, ut32 *distance, double *similarity) {
	*distance=1234;
	*similarity=1.1234;
	return true;
}
```
```
# valgrind --leak-check=full radiff2 -sV nva nvb
similarity: 1.123
distance: 1234
==47798== 
==47798== HEAP SUMMARY:
==47798==     in use at exit: 26,456 bytes in 189 blocks
==47798==   total heap usage: 279 allocs, 90 frees, 50,694 bytes allocated
~~
==47798== LEAK SUMMARY:
==47798==    definitely lost: 80 bytes in 1 blocks
==47798==    indirectly lost: 68 bytes in 2 blocks
==47798==      possibly lost: 0 bytes in 0 blocks
==47798==    still reachable: 0 bytes in 0 blocks
==47798==         suppressed: 26,308 bytes in 186 blocks
```

vs
```

==49326== HEAP SUMMARY:
==49326==     in use at exit: 26,464 bytes in 188 blocks
==49326==   total heap usage: 281 allocs, 93 frees, 80,942 bytes allocated
~~
==49326== LEAK SUMMARY:
==49326==    definitely lost: 80 bytes in 1 blocks
==49326==    indirectly lost: 68 bytes in 2 blocks
==49326==      possibly lost: 0 bytes in 0 blocks
==49326==    still reachable: 0 bytes in 0 blocks
==49326==         suppressed: 26,316 bytes in 185 blocks
```

I don't have time to chase this up, no leak in ordinary non statistical diff though, if that provides any clues.

Thanks guys!

**Comment by NikolaiHampton** [comment]
Sorry all, I'm out of time on this, I have to get back to my research papers. If you want to merge/close/ignore this it's fine by me (I have it doing what I need for my experiments). Otherwise, I'll try to look at it again in a few weeks.

**Comment by NikolaiHampton** [comment]
Sorry alvarofe I didn't see your message.. I see the heap buffer overflow I'll fix that now.

**Comment by NikolaiHampton** [comment]
Sorry @alvarofe I didn't see your message.. I see the heap buffer overflow I'll fix that now.

**Comment by NikolaiHampton** [comment]
Thanks @alvarofe - Looks like I had an old variable name in there.. Good tip on the ASAN too, I will compile everything with ASAN for testing purposes!

**Comment by Maijin** [comment]
You can use `./sys/asan.sh`

**Comment by NikolaiHampton** [comment]
@alvarofe The heap buffer overflow has been resolved. Thanks.

**Comment by alvarofe** [comment]
Still get read of size 1

```
air:r2 alvaro$ radiff2 -g main /usr/bin/false /usr/bin/true
=================================================================
==78956==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200004bdfb at pc 0x00010e59fcf2 bp 0x7fff53df3880 sp 0x7fff53df3878
READ of size 1 at 0x60200004bdfb thread T0
    #0 0x10e59fcf1 in r_diff_buffers_distance diff.c:335
    #1 0x10c4ad82f in r_anal_diff_fcn diff.c:214
    #2 0x10c02083b in r_core_gdiff gdiff.c:54
    #3 0x10be0d154 in main radiff2.c:422
    #4 0x7fff9b1db5ac in start (libdyld.dylib+0x35ac)
    #5 0x4  (<unknown module>)

ASAN:DEADLYSIGNAL
==78956==AddressSanitizer: while reporting a bug found another one. Ignoring.

air:r2 alvaro$ radiff2 -A -C /usr/bin/false /usr/bin/true
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze len bytes of instructions for references (aar)
[x] Analyze function calls (aac)
[ ] [*] Use -AA or aaaa to perform additional experimental analysis.
[x] Constructing a function name for fcn.* and sym.func.* functions (aan))
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze len bytes of instructions for references (aar)
[x] Analyze function calls (aac)
[ ] [*] Use -AA or aaaa to perform additional experimental analysis.
[x] Constructing a function name for fcn.* and sym.func.* functions (aan))
=================================================================
==78981==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200004a69b at pc 0x00010f20dcf2 bp 0x7fff53176880 sp 0x7fff53176878
READ of size 1 at 0x60200004a69b thread T0
    #0 0x10f20dcf1 in r_diff_buffers_distance diff.c:335
    #1 0x10d12182f in r_anal_diff_fcn diff.c:214
    #2 0x10cca283b in r_core_gdiff gdiff.c:54
    #3 0x10ca8a253 in main radiff2.c:428
    #4 0x7fff9b1db5ac in start (libdyld.dylib+0x35ac)
    #5 0x4  (<unknown module>)

ASAN:DEADLYSIGNAL
==78981==AddressSanitizer: while reporting a bug found another one. Ignoring.
```

The line is 

`cost = (aBufPtr[i] == bBufPtr[j]) ? 0 : 1;`

However with `radiff2 -g main /bin/ls /usr/bin/true`  i don't get any crash

**Comment by alvarofe** [comment]
```
air:r2 alvaro$ r2 -v
radare2 0.10.5-git 11945 @ darwin-x86-64 git.0.10.3-469-gd9afecf
commit: d9afecf24b4a45a270902924ac381a74b3d9faf4 build: 2016-07-14
air:r2 alvaro$ git hist -3
*   N d9afecf 2016-07-14 | Merge branch 'master' of https://github.com/NikolaiHampton/radare2 into NikolaiHampton-master (HEAD -> NikolaiHampton-master) [Álvaro Felipe Melchor]
|\
| * N 5829fd6 2016-06-21 | Optimisation of Levenshtein calc in radiff2 -s [NikolaiHampton]
* | N cb0280d 2016-07-13 | Fix crash in ec command (upstream/master, master) [Álvaro Felipe Melchor]
```

**Comment by NikolaiHampton** [comment]
Hmm.. Yeah I don't know. The original buffer overflow was from a screwup, using the wrong pointer. This one is in the compare. I can have a look in a few weeks when I'm done with some other work. Possible it's an off-by-one at the end of the compare loop? Don't know.

**Comment by alvarofe** [comment]
yeah, it was an off-by-one fixed myself and merged into master. Thx for the improvements :)

**Comment by radare** [comment]
👍
